tools/nxstyle.c

```diff
@@ -831,6 +831,57 @@ int main(int argc, char **argv, char **envp)
                ERROR("Missing file header comment block", lineno, 1);
             }
 
+          if (lineno == 2)
+            {
+              if (line[n] == '*' && line[n + 1] == '\n')
+                {
+                  ERROR("Missing relative file path in file header", lineno,
+                        n);
+                }
+              else if (isspace(line[n + 2]))
+                {
+                  ERROR("Too many whitespaces before relative file path",
+                        lineno, n);
+                }
+              else
+                {
+                  bool file_header_mismatch = false;
+
+                  /* Checking whether g_file_name has the absolute path to
+                   * the file.
+                   * NOTE: Assuming that the project root folder name is
+                   *       "nuttx".
+                   */
+
+                  const char *root_substr = "/nuttx/";
+                  char *basedir = strstr(g_file_name, root_substr);
```

Comment:
Doesn't this assume that the name of the root directory is nuttx/?  That is not always the case.

Reply:
Yes, it is one of the limitations I listed in the description.
I haven't found a nice way of making nxstyle handle both relative and absolute file paths, since the program has no knowledge about the environment.
One way would be to provide the project root path via program arguments, but I think it is a bad solution.

Reply:
how about this:
```
const char *tmp = strrstr(g_file_name, line + 2);
if (tmp && strlen(tmp) == strlen(line + 2) &&
    (tmp == g_file_name || *(tmp - 1) == '/'))
```

Reply:
`strrstr` seems to be a C extension specific to Linux. It is not available in macOS.
We could implement it ourselves in nxstyle, though.

Reply:
If so we can change to:
```
ssize_t base = strlen(g_file_name) - strlen(line + 2);
if (base >= 0 && (base == 0 || g_file_name[base - 1] == '/') &&
   strcmp(&g_file_name[base], line + 2) == 0
```

Reply:
I've just pushed a new proposal. Now `nxstyle` will only verify the relative file path if it knows the actually project root folder.
The project root folder will be provided via Makefile, which is always the case when using via `checkpatch.sh`.

Reply:
> If so we can change to:
> 
> ```
> ssize_t base = strlen(g_file_name) - strlen(line + 2);
> if (base >= 0 && (base == 0 || g_file_name[base - 1] == '/') &&
>    strcmp(&g_file_name[base], line + 2) == 0
> ```

There are two issues with your suggested implementation:
1) `line` contains the `\n` char at the end, this can be easily handled by calculating `base` with:
`ssize_t base = strlen(g_file_name) - (strlen(&line[n + 2]) - 1);`
2) `base` seems to assume that the provided relative path in the header has at least has the same amount of characters. But this is usually not the case. If you remove a single char from the `tools/nxstyle.c` at the header, it will fail to recognize the issue.

Reply:
> 2. `base` seems to assume that the provided relative path in the header has at least has the same amount of characters. But this is usually not the case. If you remove a single char from the `tools/nxstyle.c` at the header, it will fail to identify the issue.

Why not? "g_file_name[base - 1] == '/"' should catch it.

Reply:
Suppose the relative file path in the comment if `tools/nxstle.c`
This condition will not be satisfied because `g_file_name[base - 1]` will point to the 't' character.

Reply:
> Suppose the relative file path in the comment is `tools/nxstle.c`

If user forget 'y' after 't', strcmp will report the mismatch.

> This condition will not be satisfied because `g_file_name[base - 1]` will point to the 't' character.

If user typo 'ools/nxstyle.c', g_file_name[base] equals 'o' and g_file_name[base - 1] equals 't'. Then "g_file_name[base - 1] == '/"' will catch it.

Reply:
Ok, now I get. At first I thought your suggestion's objective was to identify the error, but now I understand that the condition you created will identify the correct file path. So the intended outcome is this:
```c
    if (base >= 0 && (base == 0 || g_file_name[base - 1] == '/') &&
        strcmp(&g_file_name[base], &line[n + 2]) == 0)
        {
            printf("No error\n");
        }
    else
        {
            printf("Error detected\n");
        }
```
Yeah, this seems to work well. But without the knowledge of the project's root path, it will fail to report an issue when the first directory is missing:

https://godbolt.org/z/6Ec7oq

Reply:
Yes, you are right: my algo can't catch all error, but the frequent error.